modin/core/dataframe/pandas/dataframe/utils.py

```diff
@@ -0,0 +1,349 @@
+    #     )
+
+    sample_fn = terasort_sample_fn
+
+    def pivot_fn(samples):
+        key = kwargs.get("key", None)
+        return pick_pivots_from_quantiles_for_sort(modin_frame, samples, method, key)
+
+    def split_fn(partition, pivots):
+        return split_partitions_using_pivots_for_sort(
+            modin_frame, partition, columns, pivots, ascending, **kwargs
+        )
+
+    return {
+        "sample_function": sample_fn,
+        "pivot_function": pivot_fn,
+        "split_function": split_fn,
+    }
+
+
+def _find_quantiles(
+    df: Union[pandas.DataFrame, pandas.Series], quantiles: list, method: str
+) -> np.ndarray:
+    """
+    Find quantiles of a given dataframe using the specified method.
+
+    We use this method to provide backwards compatibility with NumPy versions < 1.23 (e.g. when
+    the user is using Modin in compat mode). This is basically a wrapper around `np.quantile` that
+    ensures we provide the correct `method` argument - i.e. if we are dealing with objects (which
+    may or may not support algebra), we do not want to use a method to find quantiles that will
+    involve algebra operations (e.g. mean) between the objects, since that may fail.
+
+    Parameters
+    ----------
+    df : pandas.DataFrame or pandas.Series
+        The data to pick quantiles from.
+    quantiles : list[float]
+        The quantiles to compute.
+    method : str
+        The method to use. `linear` if dealing with numeric types, otherwise `inverted_cdf`.
+
+    Returns
+    -------
+    np.ndarray
+        A NumPy array with the quantiles of the data.
+    """
+    if method == "linear":
+        # This is the default method for finding quantiles, so it does not need to be specified,
+        # which keeps backwards compatibility with older versions of NumPy that do not have a
+        # `method` keyword argument in np.quantile.
+        return np.quantile(df, quantiles)
+    else:
+        try:
+            return np.quantile(df, quantiles, method=method)
+        except Exception:
+            # In this case, we're dealing with an array of objects, but the current version of
+            # NumPy does not have a `method` kwarg. We need to use the older kwarg, `interpolation`
+            # instead.
+            return np.quantile(df, quantiles, interpolation="lower")
+
+
+def pick_samples_for_quantiles(
+    df: pandas.DataFrame,
+    columns: list,
+    num_partitions: int,
+    length: int,
+) -> np.ndarray:
+    """
+    Pick samples over the given partition.
+
+    This function picks samples from the given partition using the TeraSort algorithm - each
+    value is sampled with probability 1 / m * ln(n * t) where m = total_length / num_partitions,
+    t = num_partitions, and n = total_length.
+
+    Parameters
+    ----------
+    df : pandas.Dataframe
+        The dataframe to pick samples from.
+    columns : list[str]
+        The columns to pick quantiles from. Only the first column in the list will be used.
```

Comment:
@RehanSD we should make this function take just `df[column]` instead of the entire `df` and `column`. I don't think there's a point in passing in the rest of `df` if it will never use. Also does this function work correctly when multiple columns are named `column`?

Reply:
In order to get only the one column to this function, we would need to do a mask on the partition, which I don't think will save any time, since we would go from deserialize the whole partition, run function and serialize results to deserialize the whole partition, mask, run function, and serialize the results, so I don't think there would be any perf implications.

For the second, that's an interesting question. I tried it out in pandas and this is what happens:
```python
In [1]: import pandas as pd
import n
In [2]: df = pd.DataFrame([['a', 'b', 'cb'], ['a', 'e', 'ca']], columns=['col0', 'col1', 'col1'])

In [3]: df.sort_values('col1')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 df.sort_values('col1')

File ~/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/util/_decorators.py:331, in deprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
    325 if len(args) > num_allow_args:
    326     warnings.warn(
    327         msg.format(arguments=_format_argument_list(allow_args)),
    328         FutureWarning,
    329         stacklevel=find_stack_level(),
    330     )
--> 331 return func(*args, **kwargs)

File ~/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/frame.py:6909, in DataFrame.sort_values(self, by, axis, ascending, inplace, kind, na_position, ignore_index, key)
   6905 elif len(by):
   6906     # len(by) == 1
   6908     by = by[0]
-> 6909     k = self._get_label_or_level_values(by, axis=axis)
   6911     # need to rewrap column in Series to apply key function
   6912     if key is not None:
   6913         # error: Incompatible types in assignment (expression has type
   6914         # "Series", variable has type "ndarray")

File ~/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/generic.py:1865, in NDFrame._get_label_or_level_values(self, key, axis)
   1862         multi_message = ""
   1864     label_axis_name = "column" if axis == 0 else "index"
-> 1865     raise ValueError(
   1866         f"The {label_axis_name} label '{key}' is not unique.{multi_message}"
   1867     )
   1869 return values

ValueError: The column label 'col1' is not unique.
```
With Modin, the following happens:
```python
In [4]: import modin.pandas as pd

In [5]: df = pd.DataFrame(df)
df.sUserWarning: Ray execution environment not yet initialized. Initializing...
To remove this warning, run the following python code before doing dataframe operations:

    import ray
    ray.init(runtime_env={'env_vars': {'__MODIN_AUTOIMPORT_PANDAS__': '1'}})

ort_valuesUserWarning: Distributing <class 'pandas.core.frame.DataFrame'> object. This may take some time.

In [6]: df.sort_values('col1')
---------------------------------------------------------------------------
RayTaskError(ValueError)                  Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 df.sort_values('col1')

File ~/Documents/modin/modin/logging/logger_decorator.py:128, in enable_logging.<locals>.decorator.<locals>.run_and_log(*args, **kwargs)
    113 """
    114 Compute function with logging if Modin logging is enabled.
    115
   (...)
    125 Any
    126 """
    127 if LogMode.get() == "disable":
--> 128     return obj(*args, **kwargs)
    130 logger = get_logger()
    131 logger_level = getattr(logger, log_level)

File ~/Documents/modin/modin/pandas/base.py:2706, in BasePandasDataset.sort_values(self, by, axis, ascending, inplace, kind, na_position, ignore_index, key)
   2704 ascending = self._validate_ascending(ascending)
   2705 if axis == 0:
-> 2706     result = self._query_compiler.sort_rows_by_column_values(
   2707         by,
   2708         ascending=ascending,
   2709         kind=kind,
   2710         na_position=na_position,
   2711         ignore_index=ignore_index,
   2712         key=key,
   2713     )
   2714 else:
   2715     result = self._query_compiler.sort_columns_by_row_values(
   2716         by,
   2717         ascending=ascending,
   (...)
   2721         key=key,
   2722     )

File ~/Documents/modin/modin/logging/logger_decorator.py:128, in enable_logging.<locals>.decorator.<locals>.run_and_log(*args, **kwargs)
    113 """
    114 Compute function with logging if Modin logging is enabled.
    115
   (...)
    125 Any
    126 """
    127 if LogMode.get() == "disable":
--> 128     return obj(*args, **kwargs)
    130 logger = get_logger()
    131 logger_level = getattr(logger, log_level)

File ~/Documents/modin/modin/core/storage_formats/pandas/query_compiler.py:3259, in PandasQueryCompiler.sort_rows_by_column_values(self, columns, ascending, **kwargs)
   3257 if Engine.get() in ["Ray", "Dask"]:
   3258     kwargs.pop("ascending", False)
-> 3259     new_modin_frame = self._modin_frame.sort_by(
   3260         0, columns, ascending=ascending, **kwargs
   3261     )
   3262     return self.__constructor__(new_modin_frame)
   3263 ignore_index = kwargs.get("ignore_index", False)

File ~/Documents/modin/modin/logging/logger_decorator.py:128, in enable_logging.<locals>.decorator.<locals>.run_and_log(*args, **kwargs)
    113 """
    114 Compute function with logging if Modin logging is enabled.
    115
   (...)
    125 Any
    126 """
    127 if LogMode.get() == "disable":
--> 128     return obj(*args, **kwargs)
    130 logger = get_logger()
    131 logger_level = getattr(logger, log_level)

File ~/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py:126, in lazy_metadata_decorator.<locals>.decorator.<locals>.run_f_on_minimally_updated_metadata(self, *args, **kwargs)
    124     elif apply_axis == "rows":
    125         obj._propagate_index_objs(axis=0)
--> 126 result = f(self, *args, **kwargs)
    127 if apply_axis is None and not transpose:
    128     result._deferred_index = self._deferred_index

File ~/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py:2020, in PandasDataframe.sort_by(self, axis, columns, ascending, **kwargs)
   2017 # If this df only has one row partition, we don't want to do a shuffle and sort - we can
   2018 # just do a full-axis sort.
   2019 if len(self._partitions) == 1:
-> 2020     return self.apply_full_axis(
   2021         1,
   2022         sort_function,
   2023     )
   2024 if self.dtypes[columns[0]] == object:
   2025     # This means we are not sorting numbers, so we need our quantiles to not try
   2026     # arithmetic on the values.
   2027     method = "inverted_cdf"

File ~/Documents/modin/modin/logging/logger_decorator.py:128, in enable_logging.<locals>.decorator.<locals>.run_and_log(*args, **kwargs)
    113 """
    114 Compute function with logging if Modin logging is enabled.
    115
   (...)
    125 Any
    126 """
    127 if LogMode.get() == "disable":
--> 128     return obj(*args, **kwargs)
    130 logger = get_logger()
    131 logger_level = getattr(logger, log_level)

File ~/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py:126, in lazy_metadata_decorator.<locals>.decorator.<locals>.run_f_on_minimally_updated_metadata(self, *args, **kwargs)
    124     elif apply_axis == "rows":
    125         obj._propagate_index_objs(axis=0)
--> 126 result = f(self, *args, **kwargs)
    127 if apply_axis is None and not transpose:
    128     result._deferred_index = self._deferred_index

File ~/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py:2221, in PandasDataframe.apply_full_axis(self, axis, func, new_index, new_columns, dtypes)
   2183 @lazy_metadata_decorator(apply_axis="both")
   2184 def apply_full_axis(
   2185     self,
   (...)
   2190     dtypes=None,
   2191 ):
   2192     """
   2193     Perform a function across an entire axis.
   2194
   (...)
   2219     The data shape may change as a result of the function.
   2220     """
-> 2221     return self.broadcast_apply_full_axis(
   2222         axis=axis,
   2223         func=func,
   2224         new_index=new_index,
   2225         new_columns=new_columns,
   2226         dtypes=dtypes,
   2227         other=None,
   2228     )

File ~/Documents/modin/modin/logging/logger_decorator.py:128, in enable_logging.<locals>.decorator.<locals>.run_and_log(*args, **kwargs)
    113 """
    114 Compute function with logging if Modin logging is enabled.
    115
   (...)
    125 Any
    126 """
    127 if LogMode.get() == "disable":
--> 128     return obj(*args, **kwargs)
    130 logger = get_logger()
    131 logger_level = getattr(logger, log_level)

File ~/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py:126, in lazy_metadata_decorator.<locals>.decorator.<locals>.run_f_on_minimally_updated_metadata(self, *args, **kwargs)
    124     elif apply_axis == "rows":
    125         obj._propagate_index_objs(axis=0)
--> 126 result = f(self, *args, **kwargs)
    127 if apply_axis is None and not transpose:
    128     result._deferred_index = self._deferred_index

File ~/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py:2672, in PandasDataframe.broadcast_apply_full_axis(self, axis, func, other, new_index, new_columns, apply_indices, enumerate_partitions, dtypes)
   2662 new_partitions = self._partition_mgr_cls.broadcast_axis_partitions(
   2663     axis=axis,
   2664     left=self._partitions,
   (...)
   2669     keep_partitioning=True,
   2670 )
   2671 # Index objects for new object creation. This is shorter than if..else
-> 2672 kw = self.__make_init_labels_args(new_partitions, new_index, new_columns)
   2673 if dtypes == "copy":
   2674     kw["dtypes"] = self._dtypes

File ~/Documents/modin/modin/logging/logger_decorator.py:128, in enable_logging.<locals>.decorator.<locals>.run_and_log(*args, **kwargs)
    113 """
    114 Compute function with logging if Modin logging is enabled.
    115
   (...)
    125 Any
    126 """
    127 if LogMode.get() == "disable":
--> 128     return obj(*args, **kwargs)
    130 logger = get_logger()
    131 logger_level = getattr(logger, log_level)

File ~/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py:2514, in PandasDataframe.__make_init_labels_args(self, partitions, index, columns)
   2511 def __make_init_labels_args(self, partitions, index, columns) -> dict:
   2512     kw = {}
   2513     kw["index"], kw["row_lengths"] = (
-> 2514         self._compute_axis_labels_and_lengths(0, partitions)
   2515         if index is None
   2516         else (index, None)
   2517     )
   2518     kw["columns"], kw["column_widths"] = (
   2519         self._compute_axis_labels_and_lengths(1, partitions)
   2520         if columns is None
   2521         else (columns, None)
   2522     )
   2523     return kw

File ~/Documents/modin/modin/logging/logger_decorator.py:128, in enable_logging.<locals>.decorator.<locals>.run_and_log(*args, **kwargs)
    113 """
    114 Compute function with logging if Modin logging is enabled.
    115
   (...)
    125 Any
    126 """
    127 if LogMode.get() == "disable":
--> 128     return obj(*args, **kwargs)
    130 logger = get_logger()
    131 logger_level = getattr(logger, log_level)

File ~/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py:459, in PandasDataframe._compute_axis_labels_and_lengths(self, axis, partitions)
    457 if partitions is None:
    458     partitions = self._partitions
--> 459 new_index, internal_idx = self._partition_mgr_cls.get_indices(axis, partitions)
    460 return new_index, list(map(len, internal_idx))

File ~/Documents/modin/modin/logging/logger_decorator.py:128, in enable_logging.<locals>.decorator.<locals>.run_and_log(*args, **kwargs)
    113 """
    114 Compute function with logging if Modin logging is enabled.
    115
   (...)
    125 Any
    126 """
    127 if LogMode.get() == "disable":
--> 128     return obj(*args, **kwargs)
    130 logger = get_logger()
    131 logger_level = getattr(logger, log_level)

File ~/Documents/modin/modin/core/dataframe/pandas/partitioning/partition_manager.py:882, in PandasDataframePartitionManager.get_indices(cls, axis, partitions, index_func)
    880 target = partitions.T if axis == 0 else partitions
    881 new_idx = [idx.apply(func) for idx in target[0]] if len(target) else []
--> 882 new_idx = cls.get_objects_from_partitions(new_idx)
    883 # TODO FIX INFORMATION LEAK!!!!1!!1!!
    884 total_idx = new_idx[0].append(new_idx[1:]) if new_idx else new_idx

File ~/Documents/modin/modin/logging/logger_decorator.py:128, in enable_logging.<locals>.decorator.<locals>.run_and_log(*args, **kwargs)
    113 """
    114 Compute function with logging if Modin logging is enabled.
    115
   (...)
    125 Any
    126 """
    127 if LogMode.get() == "disable":
--> 128     return obj(*args, **kwargs)
    130 logger = get_logger()
    131 logger_level = getattr(logger, log_level)

File ~/Documents/modin/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition_manager.py:117, in PandasOnRayDataframePartitionManager.get_objects_from_partitions(cls, partitions)
    113         partitions[idx] = part.force_materialization()
    114 assert all(
    115     [len(partition.list_of_blocks) == 1 for partition in partitions]
    116 ), "Implementation assumes that each partition contains a signle block."
--> 117 return RayWrapper.materialize(
    118     [partition.list_of_blocks[0] for partition in partitions]
    119 )

File ~/Documents/modin/modin/core/execution/ray/common/engine_wrapper.py:92, in RayWrapper.materialize(cls, obj_id)
     77 @classmethod
     78 def materialize(cls, obj_id):
     79     """
     80     Get the value of object from the Plasma store.
     81
   (...)
     90         Whatever was identified by `obj_id`.
     91     """
---> 92     return ray.get(obj_id)

File ~/.miniconda3/envs/modin/lib/python3.8/site-packages/ray/_private/client_mode_hook.py:105, in client_mode_hook.<locals>.wrapper(*args, **kwargs)
    103     if func.__name__ != "init" or is_client_mode_enabled_by_default:
    104         return getattr(ray, func.__name__)(*args, **kwargs)
--> 105 return func(*args, **kwargs)

File ~/.miniconda3/envs/modin/lib/python3.8/site-packages/ray/worker.py:1809, in get(object_refs, timeout)
   1807     worker.core_worker.dump_object_store_memory_usage()
   1808 if isinstance(value, RayTaskError):
-> 1809     raise value.as_instanceof_cause()
   1810 else:
   1811     raise value

RayTaskError(ValueError): ray::_apply_func() (pid=15002, ip=127.0.0.1)
  At least one of the input arguments for this task could not be computed:
ray.exceptions.RayTaskError: ray::deploy_ray_func() (pid=15002, ip=127.0.0.1)
  File "/Users/rehandurrani/Documents/modin/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py", line 560, in deploy_ray_func
    result = deployer(axis, f_to_deploy, f_args, f_kwargs, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/partitioning/axis_partition.py", line 181, in deploy_axis_func
    result = func(dataframe, *f_args, **f_kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 1579, in _tree_reduce_func
    series_result = func(df, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 2005, in sort_function
    df = df.sort_values(by=columns, ascending=ascending, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/util/_decorators.py", line 331, in wrapper
    return func(*args, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/frame.py", line 6909, in sort_values
    k = self._get_label_or_level_values(by, axis=axis)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/generic.py", line 1865, in _get_label_or_level_values
    raise ValueError(
ValueError: The column label 'col1' is not unique.

In [7]: 2022-11-30 15:03:38,871	ERROR worker.py:92 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::deploy_ray_func() (pid=15002, ip=127.0.0.1)
  File "/Users/rehandurrani/Documents/modin/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py", line 560, in deploy_ray_func
    result = deployer(axis, f_to_deploy, f_args, f_kwargs, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/partitioning/axis_partition.py", line 181, in deploy_axis_func
    result = func(dataframe, *f_args, **f_kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 1579, in _tree_reduce_func
    series_result = func(df, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 2005, in sort_function
    df = df.sort_values(by=columns, ascending=ascending, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/util/_decorators.py", line 331, in wrapper
    return func(*args, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/frame.py", line 6909, in sort_values
    k = self._get_label_or_level_values(by, axis=axis)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/generic.py", line 1865, in _get_label_or_level_values
    raise ValueError(
ValueError: The column label 'col1' is not unique.
2022-11-30 15:03:38,871	ERROR worker.py:92 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::_apply_func() (pid=15002, ip=127.0.0.1)
  At least one of the input arguments for this task could not be computed:
ray.exceptions.RayTaskError: ray::deploy_ray_func() (pid=15002, ip=127.0.0.1)
  File "/Users/rehandurrani/Documents/modin/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py", line 560, in deploy_ray_func
    result = deployer(axis, f_to_deploy, f_args, f_kwargs, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/partitioning/axis_partition.py", line 181, in deploy_axis_func
    result = func(dataframe, *f_args, **f_kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 1579, in _tree_reduce_func
    series_result = func(df, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 2005, in sort_function
    df = df.sort_values(by=columns, ascending=ascending, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/util/_decorators.py", line 331, in wrapper
    return func(*args, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/frame.py", line 6909, in sort_values
    k = self._get_label_or_level_values(by, axis=axis)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/generic.py", line 1865, in _get_label_or_level_values
    raise ValueError(
ValueError: The column label 'col1' is not unique.
2022-11-30 15:03:38,872	ERROR worker.py:92 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::_apply_func() (pid=15002, ip=127.0.0.1)
  At least one of the input arguments for this task could not be computed:
ray.exceptions.RayTaskError: ray::deploy_ray_func() (pid=15002, ip=127.0.0.1)
  File "/Users/rehandurrani/Documents/modin/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py", line 560, in deploy_ray_func
    result = deployer(axis, f_to_deploy, f_args, f_kwargs, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/partitioning/axis_partition.py", line 181, in deploy_axis_func
    result = func(dataframe, *f_args, **f_kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 1579, in _tree_reduce_func
    series_result = func(df, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 2005, in sort_function
    df = df.sort_values(by=columns, ascending=ascending, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/util/_decorators.py", line 331, in wrapper
    return func(*args, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/frame.py", line 6909, in sort_values
    k = self._get_label_or_level_values(by, axis=axis)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/generic.py", line 1865, in _get_label_or_level_values
    raise ValueError(
ValueError: The column label 'col1' is not unique.
2022-11-30 15:03:38,872	ERROR worker.py:92 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::deploy_ray_func() (pid=15002, ip=127.0.0.1)
  File "/Users/rehandurrani/Documents/modin/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py", line 560, in deploy_ray_func
    result = deployer(axis, f_to_deploy, f_args, f_kwargs, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/partitioning/axis_partition.py", line 181, in deploy_axis_func
    result = func(dataframe, *f_args, **f_kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 1579, in _tree_reduce_func
    series_result = func(df, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 2005, in sort_function
    df = df.sort_values(by=columns, ascending=ascending, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/util/_decorators.py", line 331, in wrapper
    return func(*args, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/frame.py", line 6909, in sort_values
    k = self._get_label_or_level_values(by, axis=axis)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/generic.py", line 1865, in _get_label_or_level_values
    raise ValueError(
ValueError: The column label 'col1' is not unique.
2022-11-30 15:03:38,872	ERROR worker.py:92 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::_apply_func() (pid=15002, ip=127.0.0.1)
  At least one of the input arguments for this task could not be computed:
ray.exceptions.RayTaskError: ray::deploy_ray_func() (pid=15002, ip=127.0.0.1)
  File "/Users/rehandurrani/Documents/modin/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py", line 560, in deploy_ray_func
    result = deployer(axis, f_to_deploy, f_args, f_kwargs, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/partitioning/axis_partition.py", line 181, in deploy_axis_func
    result = func(dataframe, *f_args, **f_kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 1579, in _tree_reduce_func
    series_result = func(df, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 2005, in sort_function
    df = df.sort_values(by=columns, ascending=ascending, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/util/_decorators.py", line 331, in wrapper
    return func(*args, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/frame.py", line 6909, in sort_values
    k = self._get_label_or_level_values(by, axis=axis)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/generic.py", line 1865, in _get_label_or_level_values
    raise ValueError(
ValueError: The column label 'col1' is not unique.
2022-11-30 15:03:38,872	ERROR worker.py:92 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::deploy_ray_func() (pid=15002, ip=127.0.0.1)
  File "/Users/rehandurrani/Documents/modin/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py", line 560, in deploy_ray_func
    result = deployer(axis, f_to_deploy, f_args, f_kwargs, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/partitioning/axis_partition.py", line 181, in deploy_axis_func
    result = func(dataframe, *f_args, **f_kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 1579, in _tree_reduce_func
    series_result = func(df, *args, **kwargs)
  File "/Users/rehandurrani/Documents/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 2005, in sort_function
    df = df.sort_values(by=columns, ascending=ascending, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/util/_decorators.py", line 331, in wrapper
    return func(*args, **kwargs)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/frame.py", line 6909, in sort_values
    k = self._get_label_or_level_values(by, axis=axis)
  File "/Users/rehandurrani/.miniconda3/envs/modin/lib/python3.8/site-packages/pandas/core/generic.py", line 1865, in _get_label_or_level_values
    raise ValueError(
ValueError: The column label 'col1' is not unique.
```
Do you think this is ok?

Reply:
> so I don't think there would be any perf implications.

I wasn't think about perf, just about cleanliness. `sample_fn` right now runs on the whole partition, right? Why not  pass `df[column]` to `pick_samples_for_quantiles` within `sample_fn` instead of passing `df` and `column`?

Reply:
> Do you think this is ok?

I think that's okay